### PR TITLE
Optimize sine/cosine functions (lookup table)

### PR DESCRIPTION
**Pull request by gamerforEA**
Benchmark result:
FastMath: ~675000 SPS
QuickMath: ~740000 SPS

### CODE REVIEW

**Comment by gamerforEA** [comment]
Old PR: https://github.com/llbit/chunky/pull/304

**Comment by llbit** [comment]
Thanks for updating the pull request. What kind of scene are you benchmarking on?

I tried rendering a simple scene:

FastMath: 742621 SPS
QuickMath: 739357 SPS

I will try a few more scenes.

**Comment by llbit** [comment]
Tried another scene. Scene2 has much more reflections:

FastMath: 1217261 SPS
QuickMath: 784445 SPS

**Comment by llbit** [comment]
My machine: Intel i7-6700 @ 3.4 GHz

**Comment by gamerforEA** [comment]
Hm. Interesting...
My CPU: Intel i5-3570 @ 3.4 GHz
Scene (1000 samples):
test-1000

**Comment by gamerforEA** [comment]
May be for old CPU good solution is lookup table (as the expensive computation), but for new CPU - computing?

**Comment by llbit** [comment]
I set up a [benchmark suite](https://github.com/llbit/chunky-benchmarks) with 11 scenes using different liting conditions, camera settings, and indoor/outdoor environments. The scenes should be fairly representative of typical workloads. All scenes are rendered to 1000 samples per pixel, with 4 Gb of Java heap space.

I ran the benchmark suite on an Intel i7-3820 3.6 GHz with Linux and Java 1.8.0_112. This resulted in an average gain of 2.7% SPS using your sine/cosine functions. One of the benchmark ran slower by 1.6%, but I ran the benchmarks again and that benchmark was 0.7% faster in the second run.

Running the benchmark suite on an Intel i7-6700 3.4 GHz with Windows 10 and Java 1.8.0_101 resulted in an average gain of 3.5% SPS using your sine/cosine functions, with no regessions.

Here is the benchmark data (all numbers are megasamples per second).

```
fm1 = FastMath, 1st run on Linux
fm2 = FastMath, 2nd run on Linux
qm1 = QuickMath, 1st run on Linux
qm2 = QuickMath, 2nd run on Linux
fm-w = FastMath on Windows
qm-w = FastMath on Windows

         scene       fm1       fm2      fm-w       qm1       qm2      qm-w
1   city_canal 0.4640142 0.4628692 0.5970573 0.4685418 0.4755140 0.6227838
2   city_night 0.4271975 0.4302579 0.5653491 0.4362514 0.4344697 0.5667328
3      fog_ice 0.6761554 0.6766157 0.9030416 0.6861887 0.6865715 0.9607302
4       forest 0.7374495 0.7105837 0.9905158 0.7557865 0.7556223 0.9937456
5     indoors1 0.2735426 0.2760739 0.3544049 0.2690505 0.2777093 0.3568609
6     indoors2 0.4148915 0.4006009 0.5401318 0.4292928 0.4213110 0.5678370
7  jungle_lake 0.5840205 0.5782227 0.7789830 0.5920707 0.5801621 0.7982478
8         lake 0.6917124 0.7164639 0.9881362 0.7270480 0.7219728 1.0327445
9         mesa 0.7500553 0.7723181 1.0490827 0.8223389 0.7959038 1.1339649
10     oil_rig 0.7956438 0.7922675 1.1034102 0.8079380 0.8138890 1.1477103
11   shipwreck 0.4807403 0.4894673 0.6654578 0.5001157 0.4918461 0.6783283
```

In conclusion, the pull request does seem to slightly improve performance. The scene that ran slower yesterday was very uncommon, with only reflective surfaces, so does not matter so much if only that type of scene runs slower. Looking at the render output, there is some slight difference specular highlights on water, but the discrepancy might just be caused by slow convergence.

To be completely thorough I should try running these benchmarks with the latest version of Apache Commons Math (3.6.1), as they may have optimized their implementations since version 3.2.

**Comment by llbit** [comment]
The differences I've been measuring are below 3%, which might not be statistically significant. Each time I run a benchmark, the score seems to have a variance of a few percent regardless of benchmarked version.

I ran the benchmarks using Commons Math 3.6.1, and the scores were very similar to the Commons Math 3.2 version.

I will run several iterations of the benchmark to get the confidence intervals for the benchmark scores, then I can see if the measured differences between FastMath and QuickMath are significant.

**Comment by llbit** [comment]
I ran 10 benchmark iterations using Commons Math 3.6.1, and calculated the 90% confidence intervals using a T distribution for each benchmark scene. Here are the results:

```
         scene           version      mean          sd  N          ci
1   city_canal 1.4.2-17-g8cda5cb 0.4602217 0.003763736 10 0.006899354
2   city_night 1.4.2-17-g8cda5cb 0.4380201 0.005011982 10 0.009187530
3      fog_ice 1.4.2-17-g8cda5cb 0.6704525 0.010854726 10 0.019897939
4       forest 1.4.2-17-g8cda5cb 0.7411307 0.011951441 10 0.021908341
5     indoors1 1.4.2-17-g8cda5cb 0.2801079 0.004876571 10 0.008939306
6     indoors2 1.4.2-17-g8cda5cb 0.4098749 0.006665537 10 0.012218683
7  jungle_lake 1.4.2-17-g8cda5cb 0.5796047 0.009316750 10 0.017078655
8         lake 1.4.2-17-g8cda5cb 0.7032029 0.019681351 10 0.036078139
9         mesa 1.4.2-17-g8cda5cb 0.7727361 0.014789797 10 0.027111368
10     oil_rig 1.4.2-17-g8cda5cb 0.7987862 0.005279103 10 0.009677192
11   shipwreck 1.4.2-17-g8cda5cb 0.4846962 0.007751721 10 0.014209780
```

The measured improvements for the two runs I did (on the same machine as the above 10 runs) with the QuickMath implementation in the pull request are with a few exceptions inside the confidence intervals, so it is not a significant enough difference to conclude a performance benefit. The change does seem to have a slight benefit for a few scenes, but it should give an increase in performance on a majority of my benchmark scenes. The benchmark suite may not ideally represent the average type of scene, but I tried to make the scenes cover most use cases.

It might be that the change works much better on some hardware, but it might also work worse on yet other hardware, so I think it's safest to not merge this pull request.

Thank you anyway for updating the pull request.
